languages/python/oso/polar/data_filtering.py

```diff
@@ -83,30 +84,41 @@ def parse_constraint(polar, constraint):
     field = constraint["field"]
     value = constraint["value"]
 
-    value_kind = next(iter(value))
-    value = value[value_kind]
+    if value is not None:
+        value_kind = next(iter(value))
+        value = value[value_kind]
 
-    if value_kind == "Term":
-        value = polar.host.to_python(value)
-    elif value_kind == "Ref":
-        child_field = value["field"]
-        result_id = value["result_id"]
-        value = Ref(field=child_field, result_id=result_id)
-    elif value_kind == "Field":
-        value = Field(field=value)
-    else:
-        assert False, "Unknown value kind"
+        if value_kind == "Term":
+            value = polar.host.to_python(value)
+        elif value_kind == "Ref":
+            value = Ref(result_id=value['result_id'], field=value['field'])
+        else:
+            assert False, "Unknown value kind"
 
     return Constraint(kind=kind, field=field, value=value)
 
+def partition(coll, pred):
+    yes, no = [], []
+    for x in coll:
+        (yes if pred(x) else no).append(x)
+    return (yes, no)
+
+def group_refs(cons):
+    res = {}
+    for con in cons:
+        grp = res.get(con.value.result_id)
+        if grp is None:
+            grp = []
+            res[con.value.result_id] = grp
+        grp.append(con)
+    return res
 
 def ground_constraints(polar, results, filter_plan, constraints):
-    for constraint in constraints:
-        if isinstance(constraint.value, Ref):
-            ref = constraint.value
-            constraint.value = results[ref.result_id]
-            if ref.field is not None:
-                constraint.value = [getattr(v, ref.field) for v in constraint.value]
+    refs, rest = partition(constraints, lambda c: isinstance(c.value, Ref) and c.value.result_id is not None)
+    for rid, cons in group_refs(refs).items():
+        fields = [[getattr(r, c.value.field) for c in cons] for r in results[rid]]
+        rest.append(Constraint(value=fields, kind='Ins', field=[c.field for c in cons]))
```

Comment:
this is pretty appalling ... we need to merge all the ref constraints on a given result set in order to get the intersection (ie. records satisfying all constraints) instead of the union (records satisfying any constraint)